File: 프로그래머스/2/131704. 택배상자/택배상자.cpp
#include <string>
#include <vector>
#include <stack>
#include <algorithm>

using namespace std;

int solution(vector<int> order) {
    int answer = 0;
    
    int max_val = order.size();
    stack<int> st;
    int idx = 1;
    
    for(auto& cur : order) {
        
        bool isPossible = false;
        // 큐에서 담을 수 있는 경우
        while(idx <= cur){
            if(idx == cur) {
                answer++;
                idx++;
                isPossible = true;
                break;
            }
            st.push(idx++);
        }
        
        if(isPossible)
            continue;
        
        if(!st.empty() && st.top() == cur){
            answer++;
            st.pop();
            continue;
        }else{
            break;
        }
        
    }
    
    return answer;
}

File: 프로그래머스/2/131704. 택배상자/택배상자.java
import java.util.*;
import java.util.stream.*;

class Solution {
    public int solution(int[] order) {
        int answer = 0;
        
        Deque<Integer> container = new ArrayDeque<>();
        Deque<Integer> assister = new ArrayDeque<>();
        
        // 초기화
        for(int i = 1 ; i <= order.length ; i++){
            container.offer(i);
        }
        
        // 로직
        int idx = 0;
        while(idx < order.length){
            
            // 보조 컨테이너 먼저 일치 확인
            if(!assister.isEmpty() && assister.peekLast() == order[idx]){
                idx++;
                answer++;
                assister.pollLast();
                continue;
            }
            
            // 컨테이너 확인
            while(!container.isEmpty() && container.peekFirst() != order[idx]){
                assister.offer(container.pollFirst());
            }
            
            // 일치하면 idx 넘긴다
            if(!container.isEmpty()){
                idx++;
                answer++;
                container.pollFirst();
                continue;
            }
            
            break;
            
        }
        
        return answer;
    }
}